pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -170,120 +184,198 @@ func filterTagList(ctx context.Context, client *rscmgr.TagBindingsClient, parent
+	bucketFullName := fmt.Sprintf(bucketParentPathFmt, cr.Spec.Storage.GCS.Bucket)
+	tags, err := t.getTagsToBind(ctx, bucketFullName)
+	if err != nil {
+		return err
+	}
+	if len(tags) <= 0 {
+		return nil
+	}
+
+	if err := t.tagBindingsClient.CreateTagBindings(ctx, bucketFullName, tags); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getTagClientOptions returns the tag client options adding the credentials and
+// the endpoint which will be used by the client.
+func getTagClientOptions(listers *regopclient.StorageListers, endpoint string) ([]option.ClientOption, error) {
+	cfg, err := GetConfig(listers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read GCS configuration for creating tag client: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(cfg.KeyfileData)),
+		option.WithEndpoint(endpoint),
+	}
+
+	return opts, nil
+}
+
+// getTagBindingsClient returns the client to be used for creating tag bindings to
+// the resources.
+func getTagBindingsClient(ctx context.Context, listers *regopclient.StorageListers, location string) (TagBindingsService, error) {
+	endpoint := fmt.Sprintf("https://%s-%s", location, resourceManagerHostSubPath)
+	opts, err := getTagClientOptions(listers, endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create tag binding client options: %w", err)
+	}
+
+	client, err := rscmgr.NewTagBindingsRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create tag binding client: %w", err)
+	}
+	return &tagBindingsClient{client}, nil
+}
+
+// NewTagManager creates a tagServiceManager instance. Explicit Close() must
+// be called when tag service is no longer needed to Close() all clients.
+func NewTagManager(ctx context.Context, listers *regopclient.StorageListers, region string) (TagService, error) {
+	client, err := getTagBindingsClient(ctx, listers, region)
+	if err != nil || client == nil {
+		return nil, err
+	}
+	mgr := &tagServiceManager{
+		Listers:           listers,
+		tagBindingsClient: client,
+	}
+	return mgr, nil
+}
+
+// Close the tag bindings client connection to API server.
+func (c *tagBindingsClient) Close() {
+	if err := c.TagBindingsClient.Close(); err != nil {
+		klog.Errorf("failed to close tag binding client: %v", err)
+	}
+}
+
+// Close the connections created by tag manager.
+func (t *tagServiceManager) Close() {
+	t.tagBindingsClient.Close()
+}
+
+// AddTagsToStorageBucket adds the user-defined tags in the Infrastructure resource
+// to the passed GCP bucket resource. It's wrapper around addUserTagsToStorageBucket()
+// and additionally updates status condition in image registry Config resource.
+func (t *tagServiceManager) AddTagsToStorageBucket(ctx context.Context, cr *imageregistryv1.Config) error {
+	if err := t.addTagsToStorageBucket(ctx, cr); err != nil {
+		util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionFalse,
```

Comment:
What do you think of moving the `util.UpdateCondition` calls from here into the caller (`CreateStorage` in gcs.go)? 
I ask because when reading the `CreateStorage`, when I see a call to `AddTagsToStorageBucket` I wouldn't expect this call to update operator conditions, so I think it would be more straight forward if this call returned an appropriate error, and the caller used this error to update the conditions accordingly.

Reply:
Yeah, `CreateStorage` is having calls to `UpdateCondition` and it would be uniform.
Reason I added here was because, `AddTagsToStorageBucket` is called twice in `CreateStorage` and `StorageTagged` condition is [updated](https://github.com/openshift/cluster-image-registry-operator/pull/1027/commits/f4a77b05243f1b0b715d8f19eb6e66c9e226feb1#diff-6e6f5be6182d62cd1ad9867497387a3c45de233050881e4479396cea0652d4a5R374-R390) for three different reasons and thought to handle it in `AddTagsToStorageBucket`.

I could define a new method `UpdateTagCondition` in `gcs` package to avoid code repetition or add a comment at `AddTagsToStorageBucket` invoked places saying `UpdateCondition` is handled by the method. Please let me know WDYT.

Reply:
I like the idea of creating an `UpdateTagCondition` function in the `gcs` package. Since it'll only be used in the `gcs` package we should make internal, so it'd be `updateTagCondition` instead. Then you can call that from `CreateStorage`.
How does that sound?

Reply:
Yeah sounds good, I will make the required changes.